opencast-backend/annotation-impl/src/main/java/org/opencast/annotation/impl/persistence/ExtendedAnnotationServiceJpaImpl.java

```diff
@@ -124,17 +141,29 @@ public void setSearchService(SearchService searchService) {
 
   /**
    * Run <code>f</code> inside a transaction with exception handling applied.
-   *
-   * @see #exhandler
    */
-  private <A> A tx(Function<EntityManager, A> f) {
-    return persistenceEnv.<A> tx().rethrow(exhandler).apply(f);
+  private <A> A tx(java.util.function.Function<EntityManager, A> f) {
+    try {
+      return db.execTx(f);
+    } catch (Exception e) {
+      if (e instanceof ExtendedAnnotationException) {
+        throw (ExtendedAnnotationException) e;
+      } else if (e instanceof RollbackException) {
+        final Throwable cause = e.getCause();
+        String message = cause.getMessage().toLowerCase();
+        if (message.contains("unique") || message.contains("duplicate"))
+          throw new ExtendedAnnotationException(Cause.DUPLICATE);
+      } else if (e instanceof NoResultException) {
+        throw new ExtendedAnnotationException(Cause.NOT_FOUND);
+      }
+      throw new ExtendedAnnotationException(Cause.SERVER_ERROR, e);
+    }
```

Comment:
When you "defunctionalize" the exception handling like this, the following seems more idiomatic:

```suggestion
    try {
      return db.execTx(f);
    } catch (NoResultException e) {
      throw new ExtendedAnnotationException(Cause.NOT_FOUND);
    } catch (RollbackException e) {
      Throwable cause = e.getCause();
      String message = cause.getMessage().toLowerCase();
      if (message.contains("unique") || message.contains("duplicate")) {
        throw new ExtendedAnnotationException(Cause.DUPLICATE);
      }
      throw new ExtendedAnnotationException(Cause.SERVER_ERROR, e);
    } catch (RuntimeException e) {
      throw new ExtendedAnnotationException(Cause.SERVER_ERROR, e);
    }
```